File: swift/Sources/MoneyHero/Domain/MarketCacheModels.swift
public enum CacheKind: String, Codable, Equatable, Sendable {
    case marketQuote = "market-quote"
    case historicalPrices = "historical-prices"
    case currencyRates = "currency-rates"
    case goldQuote = "gold-quote"
    case goldHistory = "gold-history"
    case symbolWidget = "symbol-widget"
}

public enum CacheStatus: String, Codable, Equatable, Sendable {
    case fresh
    case stale
    case expired
    case error
}

public enum CacheSource: String, Codable, Equatable, Sendable {
    case network
    case cache
}

public struct LocalCacheMetadata: Codable, Equatable, Sendable {
    public let key: String
    public let kind: CacheKind
    public let provider: MarketDataProvider
    public let symbol: String?
    public let fetchedAt: Int
    public let staleAt: Int
    public let expiresAt: Int
    public let writtenAt: Int
    public let status: CacheStatus
    public let source: CacheSource
    public let schemaVersion: Int
    public let etag: String?
    public let error: String?

    public init(
        key: String,
        kind: CacheKind,
        provider: MarketDataProvider,
        symbol: String?,
        fetchedAt: Int,
        staleAt: Int,
        expiresAt: Int,
        writtenAt: Int,
        status: CacheStatus,
        source: CacheSource,
        schemaVersion: Int,
        etag: String?,
        error: String?
    ) {
        self.key = key
        self.kind = kind
        self.provider = provider
        self.symbol = symbol
        self.fetchedAt = fetchedAt
        self.staleAt = staleAt
        self.expiresAt = expiresAt
        self.writtenAt = writtenAt
        self.status = status
        self.source = source
        self.schemaVersion = schemaVersion
        self.etag = etag
        self.error = error
    }
}

public struct CachedMarketQuote: Codable, Equatable, Sendable {
    public let metadata: LocalCacheMetadata
    public let data: MarketQuote

    public init(metadata: LocalCacheMetadata, data: MarketQuote) {
        self.metadata = metadata
        self.data = data
    }
}

public struct CachedHistoricalPrices: Codable, Equatable, Sendable {
    public let metadata: LocalCacheMetadata
    public let data: HistoricalPriceSeries

    public init(metadata: LocalCacheMetadata, data: HistoricalPriceSeries) {
        self.metadata = metadata
        self.data = data
    }
}

File: flutter/lib/domain/market_cache_models.dart
import "market_history_models.dart";
import "market_provider_models.dart";
import "market_quote_models.dart";

enum CacheKind {
  marketQuote("market-quote"),
  historicalPrices("historical-prices"),
  currencyRates("currency-rates"),
  goldQuote("gold-quote"),
  goldHistory("gold-history"),
  symbolWidget("symbol-widget");

  const CacheKind(this.value);
  final String value;
}

enum CacheStatus {
  fresh("fresh"),
  stale("stale"),
  expired("expired"),
  error("error");

  const CacheStatus(this.value);
  final String value;
}

enum CacheSource {
  network("network"),
  cache("cache");

  const CacheSource(this.value);
  final String value;
}

class LocalCacheMetadata {
  const LocalCacheMetadata({
    required this.key,
    required this.kind,
    required this.provider,
    required this.symbol,
    required this.fetchedAt,
    required this.staleAt,
    required this.expiresAt,
    required this.writtenAt,
    required this.status,
    required this.source,
    required this.schemaVersion,
    required this.etag,
    required this.error,
  });

  final String key;
  final CacheKind kind;
  final MarketDataProvider provider;
  final String? symbol;
  final int fetchedAt;
  final int staleAt;
  final int expiresAt;
  final int writtenAt;
  final CacheStatus status;
  final CacheSource source;
  final int schemaVersion;
  final String? etag;
  final String? error;

  LocalCacheMetadata copyWith({
    CacheStatus? status,
    CacheSource? source,
    String? error,
  }) {
    return LocalCacheMetadata(
      key: key,
      kind: kind,
      provider: provider,
      symbol: symbol,
      fetchedAt: fetchedAt,
      staleAt: staleAt,
      expiresAt: expiresAt,
      writtenAt: writtenAt,
      status: status ?? this.status,
      source: source ?? this.source,
      schemaVersion: schemaVersion,
      etag: etag,
      error: error ?? this.error,
    );
  }

  Map<String, Object?> toJson() => <String, Object?>{
        "key": key,
        "kind": kind.value,
        "provider": provider.name,
        "symbol": symbol,
        "fetchedAt": fetchedAt,
        "staleAt": staleAt,
        "expiresAt": expiresAt,
        "writtenAt": writtenAt,
        "status": status.value,
        "source": source.value,
        "schemaVersion": schemaVersion,
        "etag": etag,
        "error": error,
      };

  factory LocalCacheMetadata.fromJson(Map<String, dynamic> json) {
    return LocalCacheMetadata(
      key: json["key"] as String,
      kind: CacheKind.values.firstWhere((kind) => kind.value == json["kind"]),
      provider: MarketDataProvider.values.firstWhere(
        (provider) => provider.name == json["provider"],
      ),
      symbol: json["symbol"] as String?,
      fetchedAt: json["fetchedAt"] as int,
      staleAt: json["staleAt"] as int,
      expiresAt: json["expiresAt"] as int,
      writtenAt: json["writtenAt"] as int,
      status: CacheStatus.values.firstWhere(
        (status) => status.value == json["status"],
      ),
      source: CacheSource.values.firstWhere(
        (source) => source.value == json["source"],
      ),
      schemaVersion: json["schemaVersion"] as int,
      etag: json["etag"] as String?,
      error: json["error"] as String?,
    );
  }
}

class CachedMarketQuote {
  const CachedMarketQuote({required this.metadata, required this.data});

  final LocalCacheMetadata metadata;
  final MarketQuote data;

  Map<String, Object?> toJson() => <String, Object?>{
        "metadata": metadata.toJson(),
        "data": data.toJson(),
      };

  factory CachedMarketQuote.fromJson(Map<String, dynamic> json) {
    return CachedMarketQuote(
      metadata: LocalCacheMetadata.fromJson(json["metadata"] as Map<String, dynamic>),
      data: MarketQuote.fromJson(json["data"] as Map<String, dynamic>),
    );
  }
}

class CachedHistoricalPrices {
  const CachedHistoricalPrices({required this.metadata, required this.data});

  final LocalCacheMetadata metadata;
  final HistoricalPriceSeries data;

  Map<String, Object?> toJson() => <String, Object?>{
        "metadata": metadata.toJson(),
        "data": data.toJson(),
      };

  factory CachedHistoricalPrices.fromJson(Map<String, dynamic> json) {
    return CachedHistoricalPrices(
      metadata: LocalCacheMetadata.fromJson(json["metadata"] as Map<String, dynamic>),
      data: HistoricalPriceSeries.fromJson(json["data"] as Map<String, dynamic>),
    );
  }
}
